src/ess/reduce/time_of_flight/toa_to_tof.py

```diff
@@ -0,0 +1,520 @@
+    # This should be simple enough as we are making the bins linspace, so computing
+    # bin indices is fast.
+    f = RegularGridInterpolator(
+        (
+            lookup.coords["toa"].to(unit=elem_unit(toas), copy=False).values,
+            lookup.coords["distance"].to(unit=ltotal.unit, copy=False).values,
+        ),
+        lookup.data.to(unit=elem_unit(toas), copy=False).values.T,
+        method="linear",
+        bounds_error=False,
+    )
+
+    if da.bins is not None:
+        ltotal = sc.bins_like(toas, ltotal).bins.constituents["data"]
+        toas = toas.bins.constituents["data"]
+
+    tofs = sc.array(
+        dims=toas.dims, values=f((toas.values, ltotal.values)), unit=elem_unit(toas)
+    )
+
+    out = da.copy(deep=False)
+    if out.bins is not None:
+        parts = out.bins.constituents
+        out.data = sc.bins(**parts)
+        parts["data"] = tofs
+        out.bins.coords["tof"] = _bins_no_validate(**parts)
+    else:
+        out.coords["tof"] = tofs
+    return TofData(out)
+
+
+def re_histogram_tof_data(da: TofData) -> ReHistogrammedTofData:
+    """
+    Histogrammed data that has been converted to `tof` will typically have
+    unsorted bin edges (due to either wrapping of `time_of_flight` or wavelength
+    overlap between subframes).
+    This function re-histograms the data to ensure that the bin edges are sorted.
+    It makes use of the ``to_events`` helper which generates a number of events in each
+    bin with a uniform distribution. The new events are then histogrammed using a set of
+    sorted bin edges.
+
+    WARNING:
+    This function is highly experimental, has limitations and should be used with
+    caution. It is a workaround to the issue that rebinning data with unsorted bin
+    edges is not supported in scipp.
+    We also do not support variances on the data.
+    As such, this function is not part of the default set of providers, and needs to be
+    inserted manually into the workflow.
+
+    Parameters
+    ----------
+    da:
+        TofData with the time-of-flight coordinate.
+    """
+    events = to_events(da.rename_dims(time_of_flight="tof"), "event")
+
+    # Define a new bin width, close to the original bin width.
+    # TODO: this could be a workflow parameter
+    coord = da.coords["tof"]
+    bin_width = (coord["time_of_flight", 1:] - coord["time_of_flight", :-1]).nanmedian()
+    rehist = events.hist(tof=bin_width)
+    for key, var in da.coords.items():
+        if "time_of_flight" not in var.dims:
+            rehist.coords[key] = var
+    return ReHistogrammedTofData(rehist)
+
+
+def default_parameters() -> dict:
+    """
+    Default parameters of the time-of-flight workflow.
+    """
+    return {
+        PulseStride: 1,
+        PulseStrideOffset: 0,
+        DistanceResolution: sc.scalar(0.1, unit="m"),
+        TimeOfArrivalResolution: 500,
+        LookupTableRelativeErrorThreshold: 1.0e-2,
+        SimulationSeed: 1234,
+        NumberOfNeutrons: 1_000_000,
```

Comment:
I am guessing, but it looks like some of the parameters are specific to the `tof` simulation? Can we split this up?